Lists and Keys Translation (#22)
reactjs/es.reactjs.org: content/docs/lists-and-keys.md

---
id: lists-and-keys
title: Listas y keys
permalink: docs/lists-and-keys.html
prev: conditional-rendering.html
next: forms.html
---

Primero, vamos a revisar como transformas listas en Javascript. 

Dado el código de abajo, usamos la función [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) para tomar un array de `numbers` y duplicar sus valores. Asignamos el nuevo array devuelto por `map()` a la variable `doubled` y la mostramos:

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```

Este código muestra `[2, 4, 6, 8, 10]` a la consola.

> En react, transformar arrays en listas de [elementos](/docs/rendering-elements.html) es casi idéntico.

Reviewer: ```suggestion
En React, transformar arrays en listas de [elementos](/docs/rendering-elements.html) es casi idéntico.
```

> ### Renderizando Múltiples Componentes

Reviewer: Sugerí el cambio del gerundio, que aunque su uso aquí es válido parece más un calco que la manera en que lo usaríamos en español.
```suggestion
### Renderizado de múltiples componentes
```

Author: Si, me parece bien. Suena mejor de esa forma

Primero, vamos a revisar como transformas listas en Javascript. 

Dado el código de abajo, usamos la función [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) para tomar un array de `numbers` y duplicar sus valores. Asignamos el nuevo array devuelto por `map()` a la variable `doubled` y la mostramos:

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```

> Este código muestra `[2, 4, 6, 8, 10]` a la consola.

Reviewer 2: I prefer
> **en** la consola.

> En react, transformar arrays en listas de [elementos](/docs/rendering-elements.html) es casi ídentico.

Reviewer 2: `... ídentico` is actually `... idéntico.`

);
```

[**Pruebalo en CodePen**](https://codepen.io/gaearon/pen/GjPyQr?editors=0011)

> Este código muestra una lista de numeros entre 1 y 5.

Reviewer: ```suggestion
Este código muestra una lista de números entre 1 y 5.
```

> ### Componente de Lista Básica

Reviewer: ```suggestion
### Componente básico de lista
```

);
```

> Cuando ejecutes este codigo, Seras advertido que una key debe ser proporcionada para listar items. Una "key" es un atributo especial string que debes incluir al crear listas de elementos. Vamos a discutir porque esto es importante en la proxima seccion.

Reviewer 2: An uppercase `S` made its way into `serás`.
This is how I would write the following sentence:
 > serás advertido que se debería proporcioar una key para elementos de lista.

Note how the original doc says that we **should**, which is not the same as saying that we **must** use keys, hence the `debería`, and that it also says `list items` rather that `listing items`, hence using `elementos de lista` instead of `para listar elementos`.

What do you think?

Reviewer 2: Following up on my last comment, 
> ... es un atributo especial string que ...
is perhaps too much of a literal translation when specifying the type of "key".
If write this instead:
> ... es un atributo especial **de tipo string** que ...

And finally, this is just a personal opinion, but I think `Discutiremos` is better than `Vamos a discutir` on this specific scenario.

Author: Thanks for your suggestions. I agree with you

> Cuando ejecutes este código, serás advertido que una key debería ser proporcionada para ítems de lista. Una "key" es un atributo especial string que debes incluir al crear listas de elementos. Vamos a discutir porque esto es importante en la próxima sección.

Reviewer: https://dle.rae.es/?id=TgJ7yhD#CdGUKug
```suggestion
Cuando ejecutes este código, serás advertido que una key debería ser proporcionada para ítems de lista. Una "key" es un atributo especial string que debes incluir al crear listas de elementos. Vamos a discutir por qué esto es importante en la próxima sección.
```

);
```

> [**Pruebalo en CodePen**](https://codepen.io/gaearon/pen/jrXYRR?editors=0011)

Reviewer 2: `Pruébalo`

## Keys

> Keys ayudan a React a identificar que items han cambiado, son agregados, o son eliminados. Las keys deben ser dadas a los elementos dentro del array para darle a los elementos una identidad estable:

Reviewer 2: I believe we need an article for "Keys" in Spanish here.
`Las keys ayudan ...`

> Las keys ayudan a React a identificar que items han cambiado, son agregados, o son eliminados. Las keys deben ser dadas a los elementos dentro del array para darle a los elementos una identidad estable:

Reviewer: ```suggestion
Las keys ayudan a React a identificar que ítems han cambiado, son agregados, o son eliminados. Las keys deben ser dadas a los elementos dentro del array para darle a los elementos una identidad estable:
```

);
```

> La mejor forma de elegir una key es usar un string que idenfique unicamente a un elemento de la lista entre sus hermanos. Con mas frecuencia vas a usar IDs de tus datos como key:

Reviewer 2: Maybe `.. es usar un string ...` can be changed to `... es usando un string ...`

`únicamente`?

For the last sentence, this is my sugestion:
> A menudo usarías IDs de tus datos como key:

);
```

> Cuando no tengas IDs estables para renderizar, podrias usar el indice del item como una key como ultimo recurso:

Reviewer 2: I'd change `podrías` by `puedes` or `es posible`.
By the way, note the `í` on `podrías` and `ú` in `último`.

);
```

No recomendamos usar indices para keys si el orden de los items puede cambiar. Esto puede impactar negativamente el rendimiento y puede causar problemas con el estado del componente. Revisa el articulo de Robin Pokorny para una [explicacion en profundidad de los impactos negativos de usar un indice como key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Si eliges no asignar una key explicita a la lista de items, React por defecto usara indices como keys.

Aqui esta una [explicacion en profundidad sobre porque las keys son necesarias](/docs/reconciliation.html#recursing-on-children) si estas interesado en aprender mas.

### Extraccion de Components con Keys

> Las keys solo tienen sentido en el contexto de un array circundante.

Reviewer 2: I believe the correct translation for this is:
> Las keys solo tienen sentido en el contexto del array que las envuelve.

Author: Thanks, I didn't have I word for "surrounding" .
And I just realized that I haven't transalted Components in the title

> Por ejemplo, si [extraes](/docs/components-and-props.html#extracting-components) un componente `ListItem`, debes mantener la key en el elemnto `<ListItem />` en el array en lugar de en el elemento `<li>` en el `ListItem` en si mismo.

Reviewer 2: I think this can be better translated this way:
> ... debes mantener la key en los elementos `<ListItem />`  en el array en lugar de en los elementos `<li>` en el `ListItem` en sí.

Regardless of that, note `elemnto` before `<ListItem />`, it should be `elemento/s`
Note

Author: IMO 

> ...derberias mantener la key en los elementos `<ListItem />` en el array en lugar de en el elemento `<li>` en el `<ListItem />` en si mismo.

What do you think?

);
```

> No recomendamos usar índices para keys si el orden de los ítems puede cambiar. Esto puede impactar negativamente el rendimiento y puede causar problemas con el estado del componente. Revisa el árticulo de Robin Pokorny para una [explicación en profundidad de los impactos negativos de usar un índice como key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Si eliges no asignar una key explícita a la lista de items, React por defecto usara índices como keys.

Reviewer: ```suggestion
No recomendamos usar índices para keys si el orden de los ítems puede cambiar. Esto puede impactar negativamente el rendimiento y puede causar problemas con el estado del componente. Revisa el árticulo de Robin Pokorny para una [explicación en profundidad de los impactos negativos de usar un índice como key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Si eliges no asignar una key explícita a la lista de ítems, React por defecto usará índices como keys.
```

> No recomendamos usar índices para keys si el orden de los ítems puede cambiar. Esto puede impactar negativamente el rendimiento y puede causar problemas con el estado del componente. Revisa el árticulo de Robin Pokorny para una [explicación en profundidad de los impactos negativos de usar un índice como key](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Si eliges no asignar una key explícita a la lista de items, React por defecto usara índices como keys.

Reviewer 2: >[explicacion en profundidad de] 

can maybe be
> [explicacion en detalle de] ?

Author: Here I think `explicacion en profundidad ` is closer to the original

> Aquí esta una [explicación en profundidad sobre porque las keys son necesarias](/docs/reconciliation.html#recursing-on-children) si estas interesado en aprender más.

Reviewer: La traducción no tiene que ser 100% literal. «Hay» me suena más idiomático, pero siéntete libre de elegir entre ambas opciones. Si te quedas con la existente, recuerda ponerle acento a «está».
```suggestion
Aquí hay una [explicación en profundidad sobre por qué las keys son necesarias](/docs/reconciliation.html#recursing-on-children) si estás interesado en aprender más.
```

> Aquí esta una [explicación en profundidad sobre porque las keys son necesarias](/docs/reconciliation.html#recursing-on-children) si estas interesado en aprender más.

Reviewer 2: Perhaps `Aquí hay` instead of `Aquí está`

Author: In the original text it is `Here is`, so for me it should be `Aquí está` .
It should be `Here there is` to be translated in `Aquí hay` in my opinion. What do you think?

> ### Extracción de Componentes con Keys

Reviewer: ```suggestion
### Extracción de Componentes con keys
```

Las keys solo tienen sentido en el contexto del array que las envuelve.

> Por ejemplo, si [extraes](/docs/components-and-props.html#extracting-components) un componente `ListItem`, deberías mantener la key en los elementos `<ListItem />` en el array en lugar de en el elemento `<li>` en el `ListItem` en sí.

Reviewer: Creo que suena más claro sin cambiar el sentido.
```suggestion
Por ejemplo, si [extraes](/docs/components-and-props.html#extracting-components) un componente `ListItem`, deberías mantener la key en los elementos `<ListItem />` del array en lugar de en el elemento `<li>` del propio `ListItem`.
```

> **Ejemplo: Uso Incorrecto de Key**

Reviewer: ```suggestion
**Ejemplo: Uso incorrecto de key**
```

Author: Aca creo que deberia mantener las mayusculas ya que en toda la documentacion los titulos estan de esta forma

Reviewer: @EzequielMonforte Es la norma en inglés, pero no en español. Desafortunadamente, aún no tenemos un manual de estilo. Lo dejaremos así ahora. Una vez tengamos uno, tendremos que normalizar todos los artículos acorde a la decisión que se tome.

```javascript{4,5,14,15}
function ListItem(props) {
  const value = props.value;
  return (

>     // Mal! No es necesario especificar la key aquí:

Reviewer 2: My suggestion: ` // Mal! Aquí no es necesario especificar la key:`

function NumberList(props) {
  const numbers = props.numbers;
  const listItems = numbers.map((number) =>

>     // Mal! La key debería haber sido especificada aquí:

Reviewer 2: My suggestion: ` // Mal! Aquí la key debería haber sido especificada:`

);
```

**Ejemplo: Uso Correcto de Key**

```javascript{2,3,9,10}
function ListItem(props) {
  // Correcto! No hay necesidad de especificar la key aquí:
  return <li>{props.value}</li>;
}

function NumberList(props) {
  const numbers = props.numbers;
  const listItems = numbers.map((number) =>

>     // Correcto! La key deberia ser especificada dentro del array.

Reviewer: ```suggestion
    // Correcto! La key debería ser especificada dentro del array.
```

```javascript{2,3,9,10}
function ListItem(props) {

>   // Correcto! No hay necesidad de especificar la key aquí:

Reviewer 2: My suggestion: `// Correcto! Aquí no hay necesidad de especificar la key:`

);
```

> [**Pruebalo en CodePen**](https://codepen.io/gaearon/pen/ZXeOGM?editors=0010)

Reviewer 2: `pruébalo`?

Una buena regla es que los elementos dentro de `map()` necesitan keys.

> ### Las Keys Deben Ser Solo Únicas Entre Hermanos

Reviewer: ```suggestion
### Las Keys deben ser únicas solo entre hermanos
```

);
```

Con el ejemplo de arriba, el componente `Post` puede leer `props.id`, pero no `props.key`.

> ### Incrustando map() en JSX

Reviewer: En lo adelante utilizas integrar, ¿por qué no usarlo aquí también?
```suggestion
### Integrar map() en JSX
```

Author: Me lo sugirieron en otra linea y me olvide de corregirlo aca seguramente

}
```

> JSX permite [integrar cualquier expresión](/docs/introducing-jsx.html#embedding-expressions-in-jsx) en llaves así que podemos alinear el resultado `map()`:

Reviewer: ```suggestion
JSX permite [integrar cualquier expresión](/docs/introducing-jsx.html#embedding-expressions-in-jsx) en llaves así que podemos alinear el resultado de `map()`:
```

> JSX permite [incrustar cualquier expresión](/docs/introducing-jsx.html#embedding-expressions-in-jsx) en llaves así que podemos alinear el resultado `map()`:

Reviewer 2: Perhaps we could use `integrar` instead of `incrustar`?

Author: I'll take that

}
```

[**Pruébalo en CodePen**](https://codepen.io/gaearon/pen/BLvYrB?editors=0010)

> Algunas veces esto resulta en código mas claro, pero este estilo puede ser también abusado. Como en JavaScript, depende de ti decidir cuando vale la pena extraer una variable por legibilidad. Ten en mente que si el cuerpo de `map()` esta muy anidado, puede ser un buen momento para [extraer un componente](/docs/components-and-props.html#extracting-components).

Reviewer 2: I think `también puede ser  abusado` is better then `puede ser también abusado`.

Author: What about `puede ser abusado tambien?`

Reviewer 2: I think `abusado` ought to be first, then the rest of the phrase should follow.
